### PR TITLE
KAFKA-7388 equal sign in property value for password

### PR DESCRIPTION
**Pull request by mutdmour**
Changing limit of split to 2 values, to only split to key and value. if split and value is empty, also throw error. 

* Added test to check feature with delimiter (equal sign) in different pos
* deleted test that enforced no multiple equal signs in properties 
* added tests for password overrides with symbols

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala [hunk 1]
```diff
@@ -52,4 +52,13 @@ class CommandLineUtilsTest {
     assertEquals("Value of second property should be 'second'",props.getProperty("second.property"),"second")
   }
 
+  @Test
+  def testParseArgsWithMultipleDelimiters() {
+    val argArray = Array("first.property==first","second.property=second=", "third.property=thi=rd")
+    val props = CommandLineUtils.parseKeyValueArgs(argArray, false)
+    assertEquals("Value of first property should be '=first'",props.getProperty("first.property"),"=first")
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: nit: add a space between method arguments here and other places

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala [hunk 2]
```diff
@@ -60,11 +60,11 @@ object CommandLineUtils extends Logging {
...
-    val splits = args.map(_ split "=").filterNot(_.length == 0)
+    val splits = args.map(_.split("=", 2)).filterNot(_.length == 0)
 
     val props = new Properties
     for (a <- splits) {
-      if (a.length == 1) {
+      if (a.length == 1 || (a.length == 2 && a(1).length() == 0)) {
         if (acceptMissingValue) props.put(a(0), "")
         else throw new IllegalArgumentException(s"Missing value for key ${a(0)}")
       }
```
Comment (omkreddy) [inline core/src/main/scala/kafka/utils/CommandLineUtils.scala]: below else {} is not required as we always have length 2

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala [hunk 3]
```diff
@@ -60,11 +60,11 @@ object CommandLineUtils extends Logging {
    * Parse key-value pairs in the form key=value
    */
   def parseKeyValueArgs(args: Iterable[String], acceptMissingValue: Boolean = true): Properties = {
-    val splits = args.map(_ split "=").filterNot(_.length == 0)
+    val splits = args.map(_.split("=", 2)).filterNot(_.length == 0)
 
     val props = new Properties
     for (a <- splits) {
-      if (a.length == 1) {
+      if (a.length == 1 || (a.length == 2 && a(1).length() == 0)) {
```
Comment (omkreddy) [inline core/src/main/scala/kafka/utils/CommandLineUtils.scala]: we can use a(1).isEmpty() check.

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala (hunk 2 above)
Comment (muralimani) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: Yes, the else block is not required

**Review by muralimani — COMMENTED** [review]
The final else block is not required

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala (hunk 1 above)
Comment (mutdmour) [reply core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: added 👍

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala (hunk 3 above)
Comment (mutdmour) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: changed it 👍

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala (hunk 2 above)
Comment (mutdmour) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala]: removed and added more test cases for more edge cases

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala [hunk 4]
```diff
@@ -52,4 +59,14 @@ class CommandLineUtilsTest {
     assertEquals("Value of second property should be 'second'",props.getProperty("second.property"),"second")
   }
 
+  @Test
+  def testParseArgsWithMultipleDelimiters() {
+    val argArray = Array("first.property==first","second.property=second=", "third.property=thi=rd")
+    val props = CommandLineUtils.parseKeyValueArgs(argArray, false)
+
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: nit:  add a space between method arguments at below lines

**Review by omkreddy — COMMENTED** [review]
LGTM 

cc @ijuma

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala (hunk 4 above)
Comment (mutdmour) [reply core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: fixed for the whole file

**Comment by omkreddy** [comment]
retest this please

**Comment by omkreddy** [comment]
@rajinisivaram  Can you take a look at this? often  passwords contain an '=' character.

**Comment by mutdmour** [comment]
> retest this please

@omkreddy what would you like me to retest? for manual testing, I have run the shell command line tools with password override. build seems to be failing for other reasons.

**Comment by omkreddy** [comment]
@mutdmour  Above comment is not for you :) .  By commenting  "retest this please",  we can re-trigger the Jenkins builds/tests.

**Comment by mutdmour** [comment]
it passed ✅

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala [hunk 5]
```diff
@@ -60,19 +60,15 @@ object CommandLineUtils extends Logging {
    * Parse key-value pairs in the form key=value
```
Comment (rajinisivaram) [inline core/src/main/scala/kafka/utils/CommandLineUtils.scala:60]: nit: Can you add a comment to say that `value` may contain equals?

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala [hunk 6]
```diff
@@ -29,27 +29,44 @@ class CommandLineUtilsTest {
     CommandLineUtils.parseKeyValueArgs(argArray, false)
   }
 
+  @Test(expected = classOf[java.lang.IllegalArgumentException])
+  def testParseEmptyArgWithNoDelimiter() {
+    val argArray = Array("my.empty.property")
+    CommandLineUtils.parseKeyValueArgs(argArray, false)
```
Comment (rajinisivaram) [inline core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: nit: Perhaps name the boolean arg here, `acceptMissingValue = false`? Could do the same in the existing method above as well.

**Review by rajinisivaram — COMMENTED** [review]
@mutdmour Thanks for the PR, looks good. Left a couple of minor comments, apart from that LGTM.

File: core/src/main/scala/kafka/utils/CommandLineUtils.scala (hunk 5 above)
Comment (mutdmour) [reply core/src/main/scala/kafka/utils/CommandLineUtils.scala:60]: done

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala (hunk 6 above)
Comment (mutdmour) [reply core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: done

File: core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala [hunk 7]
```diff
@@ -26,30 +26,53 @@ class CommandLineUtilsTest {
   @Test(expected = classOf[java.lang.IllegalArgumentException])
   def testParseEmptyArg() {
     val argArray = Array("my.empty.property=")
-    CommandLineUtils.parseKeyValueArgs(argArray, false)
+    val acceptMissingValue = false;
+
+    CommandLineUtils.parseKeyValueArgs(argArray, acceptMissingValue)
```
Comment (rajinisivaram) [inline core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: @mutdmour In scala, you can write this as:
```
CommandLineUtils.parseKeyValueArgs(argArray, acceptMissingValue = false)
```
and
```
CommandLineUtils.parseKeyValueArgs(argArray, acceptMissingValue = true)
```
That makes the code more readable. Same for the other uses of `CommandLineUtils.parseKeyValueArgs` in this file with the boolean.

Comment (mutdmour) [reply core/src/test/scala/unit/kafka/utils/CommandLineUtilsTest.scala]: oh I see. that's clearer. will fix it up.

**Review by rajinisivaram — APPROVED** [review]
@mutdmour Thanks for the updates, LGTM. Merging to trunk.

**Pull request merged**